crates/re_viewer/src/misc/tensor_to_gpu.rs

```diff
@@ -0,0 +1,398 @@
+use std::borrow::Cow;
+
+use bytemuck::{allocation::pod_collect_to_vec, cast_slice, Pod};
+use egui::util::hash;
+use wgpu::TextureFormat;
+
+use re_log_types::component_types::{Tensor, TensorData};
+use re_renderer::{
+    renderer::{ColorMapper, ColormappedTexture},
+    resource_managers::{GpuTexture2DHandle, Texture2DCreationDesc},
+    RenderContext,
+};
+
+use super::caches::TensorStats;
+
+// ----------------------------------------------------------------------------
+
+/// Set up tensor for rendering on the GPU.
+///
+/// This will only upload the tensor if it isn't on the GPU already.
+///
+/// `tensor_stats` is used for determining the range of the texture.
+// TODO(emilk): allow user to specify the range in ui.
+pub fn tensor_to_gpu(
+    render_ctx: &mut RenderContext,
+    debug_name: &str,
+    tensor: &Tensor,
+    tensor_stats: &TensorStats,
+    annotations: &crate::ui::Annotations,
+) -> anyhow::Result<ColormappedTexture> {
+    crate::profile_function!(format!(
+        "meaning: {:?}, dtype: {}, shape: {:?}",
+        tensor.meaning,
+        tensor.dtype(),
+        tensor.shape()
+    ));
+
+    use re_log_types::component_types::TensorDataMeaning;
+
+    match tensor.meaning {
+        TensorDataMeaning::Unknown => {
+            color_tensor_to_gpu(render_ctx, debug_name, tensor, tensor_stats)
+        }
+        TensorDataMeaning::ClassId => {
+            class_id_tensor_to_gpu(render_ctx, debug_name, tensor, tensor_stats, annotations)
+        }
+        TensorDataMeaning::Depth => {
+            depth_tensor_to_gpu(render_ctx, debug_name, tensor, tensor_stats)
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+// Color textures:
+
+fn color_tensor_to_gpu(
+    render_ctx: &mut RenderContext,
+    debug_name: &str,
+    tensor: &Tensor,
+    tensor_stats: &TensorStats,
+) -> anyhow::Result<ColormappedTexture> {
+    let texture_handle = get_or_create_texture(render_ctx, hash(tensor.id()), || {
+        let [height, width, depth] = height_width_depth(tensor)?;
+        let (data, format) = match (depth, &tensor.data) {
+            // Use R8Unorm and R8Snorm to get filtering on the GPU:
+            (1, TensorData::U8(buf)) => (cast_slice_to_cow(buf.as_slice()), TextureFormat::R8Unorm),
+            (1, TensorData::I8(buf)) => (cast_slice_to_cow(buf), TextureFormat::R8Snorm),
+
+            // Special handling for sRGB(A) textures:
+            (3, TensorData::U8(buf)) => (
+                pad_and_cast(buf.as_slice(), 255),
```

Comment:
it's a bit unfortunate that the texture upload later will do extra padding to 256 byte aligned rows. Worth a todo note that we'll fuse that in once we use the write-belt (: